src/Amalgam/entity/Entity.cpp

```diff
@@ -896,74 +889,91 @@ void Entity::SetRoot(EvaluableNode *_code, bool allocated_with_entity_enm, Evalu
 	}
 }
 
-void Entity::SetRoot(std::string &code_string, EvaluableNodeManager::EvaluableNodeMetadataModifier metadata_modifier, std::vector<EntityWriteListener *> *write_listeners)
+void Entity::SetRoot(std::string &code_string, EvaluableNodeManager::EvaluableNodeMetadataModifier metadata_modifier,
+	std::vector<EntityWriteListener *> *write_listeners)
 {
 	EvaluableNodeReference new_code = Parser::Parse(code_string, &evaluableNodeManager);
 	SetRoot(new_code, true, metadata_modifier, write_listeners);
 }
 
-void Entity::AccumRoot(EvaluableNodeReference accum_code, bool allocated_with_entity_enm, EvaluableNodeManager::EvaluableNodeMetadataModifier metadata_modifier, std::vector<EntityWriteListener *> *write_listeners)
+void Entity::AccumRoot(EvaluableNodeReference accum_code, bool allocated_with_entity_enm,
+	EvaluableNodeManager::EvaluableNodeMetadataModifier metadata_modifier,
+	std::vector<EntityWriteListener *> *write_listeners)
 {
-	if( !(allocated_with_entity_enm && metadata_modifier == EvaluableNodeManager::ENMM_NO_CHANGE))
+	if(!(allocated_with_entity_enm && metadata_modifier == EvaluableNodeManager::ENMM_NO_CHANGE))
 		accum_code = evaluableNodeManager.DeepAllocCopy(accum_code, metadata_modifier);
 
-	bool accum_has_labels = EvaluableNodeTreeManipulation::DoesTreeContainLabels(accum_code);
+	auto [new_labels, no_label_collisions] = EvaluableNodeTreeManipulation::RetrieveLabelIndexesFromTree(accum_code);
 
 	EvaluableNode *previous_root = evaluableNodeManager.GetRootNode();
-	EvaluableNodeReference new_root = AccumulateEvaluableNodeIntoEvaluableNode(EvaluableNodeReference(previous_root, true), accum_code, &evaluableNodeManager);
 
-	//need to check if still cycle free as it may no longer be
-	EvaluableNodeManager::UpdateFlagsForNodeTree(new_root);
+	//before accumulating, check to see if flags will need to be updated
+	bool node_flags_need_update = false;
+	if(previous_root == nullptr)
+	{
+		node_flags_need_update = true;
+	}
+	else
+	{
+		//need to update node flags if new_root is cycle free, but accum_node isn't
+		if(previous_root->GetNeedCycleCheck() && (accum_code != nullptr && !accum_code->GetNeedCycleCheck()))
+			node_flags_need_update = true;
+
+		//need to update node flags if new_root is idempotent and accum_node isn't
+		if(previous_root->GetIsIdempotent() && (accum_code != nullptr && accum_code->GetIsIdempotent()))
+			node_flags_need_update = true;
```

Comment:
Is "GetIsIdempotent()" an expensive call? If so, this always gets called, even if GetNeedCyccleCheck() already set  node_flags_need_update to true.

Also, looks a bug, missing "!" in accum_node for idempotency check (unless the comment is wrong).